hash_practice/exercises.py

```diff
@@ -2,28 +2,75 @@
 def grouped_anagrams(strings):
     """ This method will return an array of arrays.
         Each subarray will have strings which are anagrams of each other
-        Time Complexity: ?
-        Space Complexity: ?
+        Time Complexity: O(n^2)
+        Space Complexity: O(n)
     """
-    pass
+    if not strings:
+        return []
+    anagrams = {}
+    for string in strings:
+        sorted_string = "".join(sorted(string))
+        if sorted_string in anagrams:
+            anagrams[sorted_string].append(string)
+        else:
+            anagrams[sorted_string] = [string]
+    return list(anagrams.values())
 
 def top_k_frequent_elements(nums, k):
     """ This method will return the k most common elements
         In the case of a tie it will select the first occuring element.
-        Time Complexity: ?
-        Space Complexity: ?
+        Time Complexity: O(n)
+        Space Complexity: O(n)
     """
-    pass
+    if not nums:
+        return []
+    counts = {}
+    for num in nums:
+        if num in counts:
+            counts[num] += 1
+        else:
+            counts[num] = 1
+    sorted_counts = sorted(counts.items(), key=lambda x: x[1], reverse=True)
+    return [x[0] for x in sorted_counts[:k]]
 
+def valid_row(row):
+    if not row:
+        return False
+    for i in range(1, 10):
+        if str(i) not in row:
+            return False

```

Comment:
The Sudoku problem is optional, so no need to revisit, but something to consider if you were wondering about the test failures: this function will return `False` if a row does not contain all the numbers 1-9. 

A row can be valid without being fully filled in yet, and when `valid_row` is called from `valid_subgrid` it looks like the `row` input is 3 items long and would always return false. We want to be sure that there are no duplicate values in a given row or subgrid. One way to do that could be to create a frequency map of the characters in the row, then check that no character other than `.` has a count greater than 1.